File: scala/src/main/scala/task2.scala
import org.apache.log4j.{Level, Logger}
import org.apache.spark.{HashPartitioner, SparkConf, SparkContext}
import org.json4s._
import org.json4s.jackson.JsonMethods._
import org.json4s.jackson.Serialization
import java.io._
import scala.collection.mutable._


object task2 {
  def main(args: Array[String]): Unit = {
    Logger.getLogger("org").setLevel(Level.ERROR)
    Logger.getLogger("akka").setLevel(Level.ERROR)

    //val review_filepath = "../data/review.json"
    //val output_filepath = "abc.json"
    //val n_partition = 4

    val review_filepath = args(0)
    val output_filepath = args(1)
    val n_partition = args(2)

    val spark = new SparkConf().setAppName("task2").setMaster("local[*]")
    val sc = new SparkContext(spark)

    def get_id(x: JValue): (String, Int) = {
      val user_id = compact(render(x \ "business_id"))
      (user_id.substring(1, user_id.length - 1), 1)
    }

    val start_time_m = System.nanoTime
    val rdd_data = sc.textFile(path = review_filepath).map(line => parse(line))
    val m_time = (System.nanoTime - start_time_m) / 1e9d

    val start_time_m2 = System.nanoTime
    val temp = rdd_data.map(get_id).reduceByKey(_ + _)
    val top10_business_m1 = temp.takeOrdered(num = 10)(Ordering[(Int, String)].on(x => (-x._2, x._1)))
    val m1_time = (System.nanoTime - start_time_m2) / 1e9d
    //println("Time: ", m1_time)
    val n_partition_m1 = temp.getNumPartitions
    val n_items_m1 = temp.mapPartitions(iter => Iterator(iter.size), true).collect()


    val start_time = System.nanoTime
    val temp_m2 = rdd_data.map(get_id).partitionBy(new HashPartitioner(n_partition.toInt)).reduceByKey(_ + _)
    val top10_business_m2 = temp_m2.takeOrdered(num = 10)(Ordering[(Int, String)].on(x => (-x._2, x._1)))
    val m2_time = (System.nanoTime - start_time) / 1e9d
    //println("Time: ", m2_time)
    val n_partition_m2 = temp_m2.getNumPartitions
    val n_items_m2 = temp_m2.mapPartitions(iter => Iterator(iter.size), true).collect()

    val default = LinkedHashMap(
      "n_partition" -> n_partition_m1,
      "n_items" -> n_items_m1,
      "exe_time" -> (m1_time+m_time))
    //println(default)

    val parameter = LinkedHashMap(
      "n_partition" -> n_partition_m2,
      "n_items" -> n_items_m2,
      "exe_time" -> (m2_time+m_time))
    //println(parameter)

    val result = LinkedHashMap(
      "default" -> default,
      "customized" -> parameter)
    //println (result)

    implicit val formats = org.json4s.DefaultFormats
    val writer = new PrintWriter(new File(output_filepath))
    writer.write(Serialization.write(result))
    writer.close()
  }
}

File: python/task2.py
import json
import time
import sys
from operator import add
from pyspark import SparkContext


class spark_timing_analysis:
    """
    Class to get timing and partition details between default method and explict mention of partition.
    """

    def __init__(self, review_filepath, output_file_path, partition):
        self.sc = SparkContext("local[*]", "task2")
        self.sc.setLogLevel("WARN")
        self.review_filepath = review_filepath
        self.output_file_path = output_file_path
        self.partition = partition

    def default_method(self):
        """
        Function to return Time taken to execute,number of partition by default and partition content.
        """
        start_time = time.time()
        rdd_data = self.sc.textFile(self.review_filepath).map(
            lambda line: json.loads(line)
        )
        business_reviews = rdd_data.map(
            lambda line: (line["business_id"], 1)
        ).reduceByKey(add)
        top_10_business = business_reviews.sortBy(lambda x: (-x[1], x[0]))
        default_method_time = time.time() - start_time
        # print("Time: " + str(default_method_time))
        n_partition = top_10_business.getNumPartitions()
        n_items = top_10_business.mapPartitions(lambda x: [sum(1 for i in x)]).collect()
        # print("Partition: " + str(n_partition))
        # print("Partition Items: " + str(n_items))
        default = {}
        default["n_partition"] = n_partition
        default["n_items"] = n_items
        default["exe_time"] = default_method_time
        return default

    def parameter_method(self):
        start_time = time.time()

        def custom_partitioner(key):
            return ord(key[:1])

        rdd_data = self.sc.textFile(self.review_filepath).map(
            lambda line: json.loads(line)
        )
        business_reviews = (
            rdd_data.map(lambda line: (line["business_id"], 1))
            .partitionBy(self.partition, custom_partitioner)
            .reduceByKey(add)
        )

        top_10_business = business_reviews.sortBy(lambda x: (-x[1], x[0]))
        opt_method_time = time.time() - start_time
        n_partition_opt = top_10_business.getNumPartitions()
        n_items_opt = top_10_business.mapPartitions(
            lambda x: [sum(1 for i in x)]
        ).collect()
        customized = {}
        customized["n_partition"] = n_partition_opt
        customized["n_items"] = n_items_opt
        customized["exe_time"] = opt_method_time
        return customized

    def write_results(self, default_result, parameter_result):
        result = {}
        result["default"] = default_result
        result["customized"] = parameter_result
        with open(self.output_file_path, "w") as json_file:
            json.dump(result, json_file)


if __name__ == "__main__":
    review_filepath = sys.argv[1]
    output_file_path = sys.argv[2]
    n_partition = int(sys.argv[3])
    obj = spark_timing_analysis(review_filepath, output_file_path, n_partition)
    default_result = obj.default_method()
    parameter_result = obj.parameter_method()
    obj.write_results(default_result, parameter_result)
